### DISCUSSION TITLE
Managing Keycloak configuration through code

### DISCUSSION THREAD

**Discussion by clifff** [keycloak/keycloak#28570] [Q&A]
What strategies do people use to manage their Keycloak deployments through code? We have four permanent deployments (production, qa, dev, etc) that we want to keep in sync, but it's been a difficult process.

Specifically, we want to manage things like:

- SAML login for the master realm
- Creation and configuration of the realm we primarily use
- Client settings, secrets and mapping configuration
- Enabling and configuring custom federation adapters

Initially we juggled realm imports and exports, but had to switch to [the ansible keycloak modules](https://docs.ansible.com/ansible/latest/collections/community/general/keycloak_realm_module.html#ansible-collections-community-general-keycloak-realm-module) once it became infeasible to destroy and reimport the realms when configuration changes.

The keycloak modules mostly work but have some sharp edge cases, and we've had to contribute some changes to make them work the way we need to. We're planning some work using token exchange and fine grained permissions and expecting that functionality not to be covered by ansible yet.

Are another teams trying to manage their Keycloak configurations through code? Curious if there are different approaches that would work better

**Comment by danielFesenmeyer** [comment]
Our team is using this terraform provider: https://registry.terraform.io/providers/mrparkers/keycloak/latest/docs
But, as the name suggests, it's limited to terraform.

It supports most keycloak functionality - it also support token exchange and fine grained authz, but I'm not sure to what extent.

**Comment by rene6502** [comment]
We use https://github.com/adorsys/keycloak-config-cli
Our configuration is created from a set of YAML file stored in the repo.
With helm templating we combine these files into one file per realm and use keycloak-config-cli to deploy the configuration to various environments, each controlled individually via helm values.
Secrets can be injected via environment variables (in our case we use Kubernetes secrets).
